File: domain/src/main/java/by/rekuts/travelagency/domain/Hotel.java
package by.rekuts.travelagency.domain;

import lombok.*;
import org.hibernate.annotations.OptimisticLockType;
import org.hibernate.annotations.OptimisticLocking;
import org.hibernate.annotations.Type;
import org.hibernate.annotations.TypeDef;

import javax.persistence.*;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;
import java.io.Serializable;
import java.math.BigDecimal;
import java.util.List;

/**
 * Class Hotel, includes fields <b>hotelId</b>, <b>name</b>, <b>stars</b>,
 * <b>website</b>, <b>latitude</b>, <b>longitude</b>, <b> features</b>
 * @author Aleh_Rekuts
 */
@Data @NoArgsConstructor
@OptimisticLocking(type = OptimisticLockType.VERSION)
@Entity
@Table
@TypeDef(
        name = "features",
        typeClass = HotelEnumArrayType.class
)
public class Hotel implements Serializable {
	@Id
	@SequenceGenerator( name = "jpaSequence", sequenceName = "gpa_sequence", allocationSize = 1)
	@GeneratedValue( strategy = GenerationType.SEQUENCE, generator = "jpaSequence")
	private int id;

	@NotNull
	@Column(nullable = false)
	private String name;

	@Min(1) @Max(5)
	private int stars;

	private String website;

	@Min(-90) @Max(90)
	private BigDecimal latitude;

	@Min(-180) @Max(180)
	private BigDecimal longitude;

    @Type(type="features")
	private List<String> features;

	@Version
	private Integer version;

	/**
	 * Inner to class Hotel enum Features. It stores the names of features of hotels in the values.
	 */
	public enum Features{
		A("swimming pool"),
		B("all inclusive"),
		C("fresh towel"),
		D("carpets on the walls"),
		E("free wifi"),
		F("close to sea"),
		G("free animators"),
		H("free beer"),
		I("air conditioning"),
		J("children room");

		private final String feature;

		Features(String feature) {
			this.feature = feature;
		}

		public String getValue() {
			return feature;
		}
	}
}

File: domain/src/main/scala/by/rekuts/travelagency/domain/Hotel.scala
package by.rekuts.travelagency.domain

import java.math.BigDecimal

import javax.persistence._
import javax.validation.constraints.{Max, Min, NotNull}
import org.hibernate.annotations.{OptimisticLockType, OptimisticLocking, Type, TypeDef}

@OptimisticLocking(`type` = OptimisticLockType.VERSION)
@Entity
@Table
@TypeDef(name = "features", typeClass = classOf[HotelEnumArrayType])
class Hotel extends Serializable {
  @Id
  @SequenceGenerator(name = "jpaSequence", sequenceName = "gpa_sequence", allocationSize = 1)
  @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "jpaSequence")
  var id: Int = _

  @NotNull
  @Column(nullable = false)
  var name: String = _

  @Min(1) @Max(5)
  var stars: Int = _

  var website: String = _

  @Min(-90) @Max(90)
  var latitude: BigDecimal = _

  @Min(-180) @Max(180)
  var longitude: BigDecimal = _

  @Type(`type` = "features")
  var features: List[String] = _

  @Version
  var version: Integer =_

}

object Features extends Enumeration {
  type Features = Value
  val A = Value("swimming pool")
  val B = Value("all inclusive")
  val C = Value("fresh towel")
  val D = Value("carpets on the walls")
  val E = Value("free wifi")
  val F = Value("close to sea")
  val G = Value("free animators")
  val H = Value("free beer")
  val I = Value("air conditioning")
  val J = Value("children room")
}
